Documentation/Evolution/RegexSyntax.md

```diff
@@ -0,0 +1,845 @@
+- `m`: Enables `^` and `$` to match against the start and end of a line rather than only the start and end of the entire string.
+- `n`: Disables the capturing behavior of `(...)` groups. Named capture groups must be used instead.
+- `s`: Changes `.` to match any character, including newlines.
+- `U`: Changes quantifiers to be reluctant by default, with the `?` specifier changing to mean greedy.
+- `x`, `xx`: Enables extended syntax mode, which allows non-semantic whitespace and end-of-line comments. See [Extended Syntax Modes](#extended-syntax-modes) for more info.
+
+#### ICU options
+
+- `w`: Enables the Unicode interpretation of word boundaries `\b`.
+
+#### Oniguruma options
+
+- `D`: Enables ASCII-only digit matching for `\d`, `\p{Digit}`, `[:digit:]`.
+- `S`: Enables ASCII-only space matching for `\s`, `\p{Space}`, `[:space:]`.
+- `W`: Enables ASCII-only word matching for `\w`, `\p{Word}`, `[:word:]`, and `\b`.
+- `P`: Enables ASCII-only for all POSIX properties (including `digit`, `space`, and `word`).
+- `y{g}`, `y{w}`: Changes the meaning of `\X`, `\y`, `\Y`. These are mutually exclusive options, with `y{g}` specifying extended grapheme cluster mode, and `y{w}` specifying word mode.
+
+#### Swift options
+
+These options are specific to the Swift regex matching engine and control the semantic level at which matching takes place.
+
+- `X`: Grapheme cluster matching.
+- `u`: Unicode scalar matching.
+- `b`: Byte matching.
+
+Further details on these are TBD and outside the scope of this pitch.
+      
+### References
+
+```
+NamedOrNumberRef -> NamedRef | NumberRef
+NamedRef         -> Identifier RecursionLevel?
+NumberRef        -> ('+' | '-')? <Decimal Number> RecursionLevel?
+RecursionLevel   -> '+' <Int> | '-' <Int>
+```
+
+A reference is an abstract identifier for a particular capturing group in a regular expression. It can either be named or numbered, and in the latter case may be specified relative to the current group. For example `-2` refers to the capture group `N - 2` where `N` is the number of the next capture group. References may refer to groups ahead of the current position e.g `+3`, or the name of a future group. These may be useful in recursive cases where the group being referenced has been matched in a prior iteration.
+
+A backreference may optionally include a recursion level in certain cases, which is a syntactic element inherited [from Oniguruma][oniguruma-syntax] that allows the reference to specify a capture relative to a given recursion level.
+
+#### Backreferences
+
+```
+Backreference -> '\g{' NamedOrNumberRef '}'
+               | '\g' NumberRef
+               | '\k<' NamedOrNumberRef '>'
+               | "\k'" NamedOrNumberRef "'"
+               | '\k{' NamedRef '}'
+               | '\' [1-9] [0-9]+
+               | '(?P=' NamedRef ')'
+```
+
+A backreference evaluates to the value last captured by the referenced capturing group. If the referenced capture has not been evaluated yet, the match fails.
```

Comment:
Should we say something like "if a referenced capture does not exist anywhere in the regular expression, the regular expression is not valid"?

Reply:
How's this?